// @keystonejs/fields-location-google/views/Field/dist/fields-location-google.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var core = require('@emotion/core');
var react = require('react');
var reactToastNotifications = require('react-toast-notifications');
var fields = require('@arch-ui/fields');
var Select = require('@arch-ui/select');
var googleMapsReact = require('google-maps-react');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var Select__default = /*#__PURE__*/_interopDefault(Select);

const LocationGoogleField = ({
  field,
  value,
  errors,
  onChange,
  google,
  renderContext,
  isDisabled
}) => {
  const {
    googlePlaceID,
    formattedAddress,
    lat,
    lng
  } = _objectSpread({}, value);

  const htmlID = `ks-input-${field.path}`;
  const autocompleteService = new google.maps.places.AutocompleteService();
  const geocoder = new google.maps.Geocoder();
  const {
    addToast
  } = reactToastNotifications.useToasts();
  const [inputValue, setInputValue] = react.useState(googlePlaceID ? {
    label: formattedAddress,
    value: googlePlaceID
  } : null);
  const [marker, setMarker] = react.useState(lat && lng ? {
    lat,
    lng
  } : null);

  const handleOptionChange = option => {
    if (!option) {
      onChange(null);
      setMarker(null);
      setInputValue(null);
      return;
    }

    const placeId = option.value;
    geocoder.geocode({
      placeId
    }, (results, status) => {
      if (status === 'OK') {
        if (results[0]) {
          const {
            formatted_address,
            geometry: {
              location: {
                lat,
                lng
              }
            }
          } = results[0];
          setInputValue({
            label: formatted_address,
            value: placeId
          });
          setMarker({
            lat: lat(),
            lng: lng()
          });
          onChange(placeId);
        }
      } else {
        addToast('Could not find the provided location.', {
          appearance: 'error',
          autoDismiss: true
        });
      }
    });
  };

  const loadOptions = inputValue => new Promise(resolve => {
    autocompleteService.getPlacePredictions({
      input: inputValue
    }, results => {
      if (results) {
        resolve(results.map(({
          description,
          place_id
        }) => ({
          label: description,
          value: place_id
        })));
      }

      resolve(null);
    });
  });

  const selectProps = renderContext === 'dialog' ? {
    menuPortalTarget: document.body,
    menuShouldBlockScroll: true
  } : null;
  return core.jsx(fields.FieldContainer, null, core.jsx(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), core.jsx(fields.FieldDescription, {
    text: field.adminDoc
  }), core.jsx(fields.FieldInput, {
    css: {
      flexDirection: 'column'
    }
  }, core.jsx(Select__default['default'], _extends({
    isAsync: true,
    isClearable: true,
    cacheOptions: true,
    placeholder: "Search for a location ...",
    value: inputValue,
    onChange: handleOptionChange,
    loadOptions: loadOptions,
    id: `react-select-${htmlID}`,
    inputId: htmlID,
    instanceId: htmlID,
    css: {
      width: '100%'
    },
    isDisabled: isDisabled
  }, selectProps)), marker && core.jsx("div", {
    css: {
      position: 'relative',
      height: '14rem',
      marginTop: '1rem'
    }
  }, core.jsx(googleMapsReact.Map, {
    google: google,
    initialCenter: marker,
    center: marker,
    zoom: 16
  }, core.jsx(googleMapsReact.Marker, {
    position: marker
  })))));
};

var Field = googleMapsReact.GoogleApiWrapper(props => ({
  apiKey: props.field.config.googleMapsKey
}))(LocationGoogleField);

exports.default = Field;
